[Woo POS] Always create a new order when checking out

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15426
Discussion: https://github.com/woocommerce/woocommerce-android/issues/12385
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Always create a new order when checking out. This matches the Android solution. Updating the order requires additional diffing logic on the mobile side and sometimes it [introduces issues when remote and local orders become out-of-sync](https://github.com/woocommerce/woocommerce-ios/issues/15426).

`auto-draft` orders [are removed by WordPress once 7 days.](https://github.com/WordPress/WordPress/blob/a645d169b2491a17d07efbf063c830c2b0e31c36/wp-includes/post.php#L8063). 

We could consider creating `checkout-draft` orders, that are [removed once a day by a cron job.](https://woocommerce.com/document/managing-orders/order-statuses/#draft-order-cleanup)

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### The original issue I was addressing 

Prerequisites: Enable `enableCouponsInPointOfSale` feature flag in `DefaultFeatureFlagService`

1. Add 2 products
3. Check out so the order would be created
4. Come back
5. Remove one of the producs
6. Add one invalid coupon
7. Check out
8. Get error
9. Come back
10. Remove an invalid coupon (the Cart should contain only one product now)
11. Check out again
12. Confirm order is created successfully

### Regression

The changes affect the checkout flow, which could be triggered in two ways:
1. Tapping Check Out
2. Tapping 'Try Again' if order creation fails

Test with various types of products:
1. Select products -> Check out -> Come back -> Make Changes -> Card Payment
2. Select products -> Check out -> Order creation failure (e.g. no network connection) -> Address Issue (Enable network) -> Try Again -> Card Payment
3. Select products -> Check out -> Come back -> Checkout -> Order is not loading since it remains unchanged

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested various scenarios with different types of products, variations, and coupons. Also, cash payment and emails. 

The order creation logic is not changed, I use the same method for adding order items to the order. Order editing works as order creation now. 

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Always create a new order when checking out

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Merge branch 'trunk' into task/15326-pos-always-create-a-new-order-during-checkout

staskus committed:
Update PointOfSaleOrderControllerTests

staskus committed:
Update RELEASE-NOTES.txt

joshheald commented:
The repro steps don't always work. When I come back at step 11, it sometimes still shows the invalid coupons error. Is that expected at this point, and unrelated to the PR? 

I'm guessing that removing the coupon isn't noticed as a change to the cart, so the order isn't recalculated... but it has worked sometimes. Adding a coupon to the cart is reliably noticed as a change, but not removing it.


https://github.com/user-attachments/assets/58cfebbd-6722-4fa0-846e-40403744eb3c


Edit: I guess that the `coupon.matches(order)` check would return true after the coupon is removed – the API removes the coupon from the order itself, right?

joshheald commented on Yosemite/Yosemite/Tools/POS/POSOrderService.swift:
```diff
@@ -66,73 +60,17 @@ public final class POSOrderService: POSOrderServiceProtocol {
-        return order
+private extension Order {
+    func addItems(_ cartItems: [POSCartItem]) -> Order {
+        let itemsToAdd = Array(cartItems.createGroupedOrderSyncProductInputs().values)
```
Do we need to call `sanitizingLocalItems` on these, or is that only relevant for updates?

staskus replied:
😨 I missed it.

Great observation. We do need it. As the comment in the `sanitizingLocalItems` says, not applying it breaks the calculations if `prices_include_tax` is `true`.

joshheald replied:
Ah, glad I spotted it then! I wasn't sure because it could have been something only required on update... though, our products are all added as an update, aren't they? Rather than in the initial POST request?

staskus replied:
> though, our products are all added as an update, aren't they? Rather than in the initial POST request?

No. We add them as initial request when creating an order. `createGroupedOrderSyncProductInputs` turns products into order items. `sanitize` makes sure totals and subtotals are cleared and don't influence backend calculations.

joshheald replied:
Ah, makes sense! Yeah, I'm all for just sending up an ID and a quantity as our input. Factors in to the line/cart-level discount/taxes discussion as well.

Ideally, if we ever allow merchants to add line-level discounts like they can in IPP, we could make a change in the API that lets us specify a discount directly, and whether that's before or after tax.

joshheald reviewed (approved):
Thanks for this – It's a nice simplification of the logic. Other than the issue mentioned about not reliably recognising invalid coupon removal, I found it worked well in my testing...

staskus committed:
Merge branch 'trunk' into task/15326-pos-always-create-a-new-order-during-checkout

staskus committed:
Sanitize order items before sending to remote

Sanitizing removes order item totals and subtotals, allowing the backend to calculate it. It's critical for correct calculations when prices are inputted with taxes.

staskus commented:
> The repro steps don't always work. When I come back at step 11, it sometimes still shows the invalid coupons error. Is that expected at this point, and unrelated to the PR?

> I'm guessing that removing the coupon isn't noticed as a change to the cart, so the order isn't recalculated... but it has worked sometimes. Adding a coupon to the cart is reliably noticed as a change, but not removing it.

Good catch. Why it happens:

1. Create and persist order (Add Product -> Check out -> Sync -> Come Back).
2. Add invalid coupon, API returns error, order doesn't change (Add Invalid Coupon -> Check out -> Sync -> Error -> Come Back).
3. Remove invalid coupon, cart state is equal to order state, sync is not initialized (Remove Coupon -> Check out -> Error).

The way to get around this cleanly is to always sync the next time after an error is returned from the API.

staskus committed:
Clean the order when syncing fails to sync again during the next checkout
